### DISCUSSION TITLE
Some of the sensors values are not refreshed

### DISCUSSION THREAD

**Discussion by tricky10** [syssi/esphome-jk-bms#606] [Q&A]
I have some issues with power, charging and other sensors which are not being updated in a loop.
The cell voltage is working as expected. I use both smg inverter and jk bms setup on my esp32. 
In logs thre is a lot of 
`[11:06:17][D][esp32_ble_client:110]: [0] [C8:47:80:1D:83:65] ESP_GATTC_NOTIFY_EVT
`

messages and there is no update for power or charing sensor why ?
I've attached my logs and confguration. Do you have any idea what I do wrong or maybe there is a bug ?

[logs_inverter_logs (3).txt](https://github.com/user-attachments/files/17276051/logs_inverter_logs.3.txt)
[jkbms-easun.txt](https://github.com/user-attachments/files/17276068/jkbms-easun.txt)

my jk-bms harder version is V11A

Thanks,
Tomasz

**Comment by syssi** [comment]
There is one `ESP_GATTC_NOTIFY_EVT` message per received frame. This is totally fine because the BMS sends 2 frames per second. Most aren't parsed/discarded because of the `throttle` setting.

Could you tell me the explicit sensor entities which doesn't get updated? Are we talking about `sensor` entities or `binary_sensor` entities?

**Reply by tricky10** [reply]
power and sometimes charing and discharging but power is the most missed. If you notice there is and update of the 
`[11:06:22][D][sensor:094]: 'jk-bms charging power': Sending state 116.66528 W with 2 decimals of accuracy
`
but only once. This should be updated every time together with this one ?


`[11:06:32][D][sensor:094]: 'jk-bms cell voltage 1': Sending state 3.32000 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 2': Sending state 3.31700 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 3': Sending state 3.32100 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 4': Sending state 3.32100 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 5': Sending state 3.31700 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 6': Sending state 3.32100 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 7': Sending state 3.32100 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 8': Sending state 3.32000 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms cell voltage 9': Sending state 0.00000 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms min cell voltage': Sending state 3.31700 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms max cell voltage': Sending state 3.32100 V with 3 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms max voltage cell': Sending state 3.00000  with 0 decimals of accuracy
[11:06:32][D][sensor:094]: 'jk-bms min voltage cell': Sending state 2.00000  with 0 decimals of accuracy`

BTW; as u can see 'jk-bms power' is not appearing at all :(.

**Reply by tricky10** [reply]
![image](https://github.com/user-attachments/assets/9400aff5-901b-4c01-ab7a-ba4cd8a5188b)

**Reply by syssi** [reply]
Are you sure the flashed version on the ESP is the recent version of your YAML configuration? It looks like the `power` sensor wasn't part of the YAML at the time you have flashed the ESP last time.

**Reply by tricky10** [reply]
It's the same YAML over and over just with debug off and on - already flashed it multiple times today. 
HINT: Maybe it's somehow related to easun config ?  I've noticed it works much better when no easun configuration is provided parrallel.

**Reply by syssi** [reply]
On heavy load the `api` components drops messages/states so they aren't passed to Home Assistant anymore. But I expect some debug messages at the ESPHome log nevertheless AFAIK. See https://github.com/esphome/esphome/pull/7321

**Reply by tricky10** [reply]
Hmmm power last update is 6 minutes ago and avg cell voltage couple of seconds ago :(
![image](https://github.com/user-attachments/assets/bbae4143-d768-4c87-92f8-fa6aba60f2f9)
![image](https://github.com/user-attachments/assets/34c9fd07-db87-4ea4-9512-1128ef5aa958)

**Reply by syssi** [reply]
If you remove all sensors (temporary) and just expose `power` does this improve the behavior?

**Reply by tricky10** [reply]
I left only 

- charging power, 
- discharging power 
- power 

and the values are constantly updated. The easun sensors are also exposed so probably there is some pefromance issue in the code or one of the sensors creates some parsing issues ? 
![image](https://github.com/user-attachments/assets/8173588b-11f0-4354-accc-a7d9da70065d)

**Reply by syssi** [reply]
You could try to switch from `api` to `mqtt` as workaround. In worst case the ESP crashs using `mqtt` because there is not enough memory to handle all the sensors.

**Reply by tricky10** [reply]
It's for sure the issue with the amount of sensors when I add all again the issue comes back

**Reply by tricky10** [reply]
I will buy another esp32-c3 and move the support for jkbms into it so I will split the management of sensors between two seperate hardwares - I will let you know if this is a code issue or rather hardware issue.

**Reply by syssi** [reply]
It's an software issue for sure and caused by https://github.com/esphome/esphome/pull/7321. The problem grows with the number of sensors!

**Reply by Kappa71** [reply]
Any news on for this problem?
I have the same issue with a PB2A16S20P via BLE.

would be a viable solution to switch to mqtt ?

**Reply by syssi** [reply]
Yes. Using MQTT is a good workaround. My knowledge is not sufficient to provide a solution here.
